clio-server/src/main/scala/org/broadinstitute/clio/server/service/IndexService.scala

```diff
@@ -67,4 +95,70 @@ abstract class IndexService[CI <: ClioIndex](
       elasticsearchIndex
     )
   }
+
+  def validateUpsert(
+    existingMetadata: Json,
+    newMetadata: clioIndex.MetadataType
+  ): Either[Exception, clioIndex.MetadataType] = {
+    val differences = JsonDiff.diff(existingMetadata, newMetadata, remember = true)
+    /* Types of differences:
+        Replace - the existing value is different from what we want to set. We ignore nulls. This will overwrite.
+        Remove - the existing value doesn't exist in the new meta data. This will not overwrite.
+        Add - the existing value doesn't exist. This will not overwrite.
+        Move - effectively an Add and Remove. This will not overwrite.
+        Copy - effectively an Add. This will not overwrite.
+     */
+    val replaceOperations = differences.ops
+      .filter(
+        op => op.isInstanceOf[Replace] && op.asInstanceOf[Replace].value != Json.Null
```

Comment:
I think we can use `flatMap` and a pattern-match to avoid the casting.
```
val replaceOperations: Seq[Replace] = differences.ops.flatMap {
  case replace: Replace => Some(replace).filter(_.value != Json.Null)
  case _ => None
}
```
And then we shouldn't need the cast below, either.